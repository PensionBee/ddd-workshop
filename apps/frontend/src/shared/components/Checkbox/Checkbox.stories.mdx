import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Checkbox } from "./Checkbox";

<Meta
  title="Shared/Checkbox/Docs"
  component={Checkbox}
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Checkbox

## Props

<ArgsTable of={Checkbox} />

## Usage & Customisation

### Default

<Canvas>
  <Story id="shared-checkbox--default" />
</Canvas>

### Round

<Canvas>
  <Story id="shared-checkbox--round" />
</Canvas>

### Disabled

<Canvas>
  <Story id="shared-checkbox--disabled" />
</Canvas>

### Error

<Canvas>
  <Story id="shared-checkbox--error" />
</Canvas>

### Default Active

<Canvas>
  <Story id="shared-checkbox--default-active" />
</Canvas>

### Round Active

<Canvas>
  <Story id="shared-checkbox--round-active" />
</Canvas>

### Disabled Active

<Canvas>
  <Story id="shared-checkbox--disabled-active" />
</Canvas>

### Error Active

<Canvas>
  <Story id="shared-checkbox--error-active" />
</Canvas>
