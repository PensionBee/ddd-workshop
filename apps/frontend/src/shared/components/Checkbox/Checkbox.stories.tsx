import { Checkbox } from "./Checkbox";
import type { ComponentMeta } from "@storybook/react";
import { useForm } from "react-hook-form";

export default {
  title: "Shared/Checkbox",
  component: Checkbox,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof Checkbox>;

export const Default = () => {
  const { control } = useForm();

  return <Checkbox name="checkbox" control={control} />;
};

export const Round = () => {
  const { control } = useForm();

  return <Checkbox name="checkbox" control={control} variant="circle" />;
};

export const Disabled = () => {
  const { control } = useForm();

  return <Checkbox name="checkbox" control={control} disabled />;
};

export const Error = () => {
  const { control } = useForm();

  return (
    <Checkbox name="checkbox" control={control} error={"An error message"} />
  );
};

export const DefaultActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return <Checkbox name="checkbox" control={control} />;
};

export const RoundActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return <Checkbox name="checkbox" control={control} variant="circle" />;
};

export const DisabledActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return <Checkbox name="checkbox" control={control} disabled />;
};

export const ErrorActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return (
    <Checkbox name="checkbox" control={control} error={"An error message"} />
  );
};
