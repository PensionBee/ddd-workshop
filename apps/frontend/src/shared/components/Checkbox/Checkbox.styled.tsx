import * as CheckboxPrimitive from "@radix-ui/react-checkbox";
import { theme } from "@shared/styling/theme";
import styled, { css } from "styled-components";
import { ComponentIcon } from "../Svg/Svg";

export type StyledCheckboxButtonProps = {
  $variant: "square" | "circle";
  $error?: string;
};

export const StyledCheckboxButton = styled(
  CheckboxPrimitive.Root
)<StyledCheckboxButtonProps>`
  position: relative;
  cursor: pointer;
  width: 22px;
  height: 22px;
  overflow: unset;
  border-radius: 4px;
  border: 2px solid ${() => theme.colors.grey3};
  &:focus,
  button:focus & {
    outline: 2px solid ${() => theme.colors.yellow};
  }
  background-color: white;
  &[aria-disabled="true"] {
    opacity: 0.4;
    cursor: not-allowed;
  }
  ${({ $error }) =>
    $error &&
    css`
      outline: 2px solid ${() => theme.colors.red100};
    `}
  ${({ $variant }) =>
    $variant === "circle" &&
    css`
      width: 23px;
      height: 23px;
      border-radius: 50%;
    `}
  & + input {
    display: none;
  }
`;

export const StyledIcon = styled(CheckboxPrimitive.Indicator)`
  display: flex;
`;

const Check = (
  props: Omit<React.ComponentProps<typeof ComponentIcon>, "icon">
) => <ComponentIcon icon="Check" {...props} />;

export const StyledCheckmark = styled(Check)<StyledCheckboxButtonProps>`
  position: absolute;
  bottom: 0px;
  left: 0px;
  [aria-pressed="true"] & path {
    fill: ${() => theme.colors.teal100};
  }
  [aria-pressed="false"] & path {
    fill: transparent;
  }
  ${({ $variant }) =>
    $variant === "circle" &&
    css`
      bottom: 1px;
      left: 1px;
    `}
`;
