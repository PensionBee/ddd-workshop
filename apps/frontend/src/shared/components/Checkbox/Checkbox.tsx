/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Control } from "react-hook-form";
import { Controller } from "react-hook-form";
import type { StyledCheckboxButtonProps } from "./Checkbox.styled";
import {
  StyledCheckboxButton,
  StyledCheckmark,
  StyledIcon,
} from "./Checkbox.styled";
import { Box } from "../Box/Box";

export type CheckboxProps = {
  /** Form control */
  control: Control<any, any>;
  /** If switch is disabled */
  disabled?: boolean;
  /** Message if error */
  error?: string;
  /** Field name */
  name: string;
  /** Input value */
  value?: React.HTMLProps<HTMLInputElement>["value"];
  /** Button variant */
  variant?: StyledCheckboxButtonProps["$variant"];
};

export const Checkbox: React.FC<CheckboxProps> = ({
  variant = "square",
  error,
  name,
  disabled,
  control,
  ...props
}) => {
  return (
    <Box d="flex" data-rc="Checkbox">
      <Controller
        control={control}
        name={name}
        render={({ field }) => (
          <StyledCheckboxButton
            key={name}
            $error={error}
            $variant={variant}
            disabled={disabled}
            aria-disabled={disabled}
            {...field}
            value={undefined}
            checked={field.value}
            onCheckedChange={field.onChange}
          >
            <StyledIcon>
              <StyledCheckmark $variant={variant} />
            </StyledIcon>
          </StyledCheckboxButton>
        )}
        {...props}
      />
    </Box>
  );
};
