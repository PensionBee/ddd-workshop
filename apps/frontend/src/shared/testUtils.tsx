import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
