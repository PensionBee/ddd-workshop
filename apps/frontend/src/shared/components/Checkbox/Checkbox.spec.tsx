import { render } from "@shared/testUtils";
import { useForm } from "react-hook-form";
import { Checkbox } from "./Checkbox";

const CHECKBOX_TEST_ID = "checkbox";
const CHECKBOX_NAME = "Checkbox name";

const CheckboxForm = () => {
  const { control } = useForm();
  return (
    <Checkbox
      data-testid={CHECKBOX_TEST_ID}
      name={CHECKBOX_NAME}
      control={control}
    />
  );
};

describe("Checkbox", () => {
  describe("Default", () => {
    it("should render correctly", () => {
      const { queryByRole, asFragment } = render(<CheckboxForm />);
      const checkbox = queryByRole("checkbox");

      expect(checkbox).toBeVisible();
      expect(checkbox).toHaveAttribute("data-state", "unchecked");
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
