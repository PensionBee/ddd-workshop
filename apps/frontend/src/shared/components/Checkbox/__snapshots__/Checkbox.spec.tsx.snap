// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-eCYdKt gJDuSr"
    data-rc="Checkbox"
  >
    <button
      aria-checked="false"
      class="sc-gsnScm dNaquf"
      data-state="unchecked"
      role="checkbox"
      type="button"
      value="on"
    />
  </div>
</DocumentFragment>
`;

exports[`Checkbox Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-eDvSVe cYuVyI"
    data-rc="Checkbox"
  >
    <button
      aria-checked="false"
      class="sc-gswNZR cSNIzy"
      data-state="unchecked"
      role="checkbox"
      type="button"
      value="on"
    />
  </div>
</DocumentFragment>
`;
